Data race in socket source

The following data race has occurred twice when running tests in CI:

```
env GO111MODULE=on go test -tags '' -race -count=1 ./...
ok  	github.com/influxdata/flux	5.346s
ok  	github.com/influxdata/flux/arrow	1.029s
ok  	github.com/influxdata/flux/ast	1.104s
?   	github.com/influxdata/flux/ast/asttest	[no test files]
?   	github.com/influxdata/flux/ast/asttest/cmpgen	[no test files]
ok  	github.com/influxdata/flux/ast/edit	1.032s
ok  	github.com/influxdata/flux/builtin	5.018s
?   	github.com/influxdata/flux/cmd/flux	[no test files]
?   	github.com/influxdata/flux/cmd/flux/cmd	[no test files]
ok  	github.com/influxdata/flux/compiler	1.100s
ok  	github.com/influxdata/flux/complete	1.039s
ok  	github.com/influxdata/flux/control	11.107s
?   	github.com/influxdata/flux/control/controltest	[no test files]
ok  	github.com/influxdata/flux/csv	1.041s
ok  	github.com/influxdata/flux/execute	5.031s
?   	github.com/influxdata/flux/execute/executetest	[no test files]
ok  	github.com/influxdata/flux/influxql	1.020s
?   	github.com/influxdata/flux/internal/cmd/builtin	[no test files]
?   	github.com/influxdata/flux/internal/cmd/builtin/cmd	[no test files]
?   	github.com/influxdata/flux/internal/cmd/changelog	[no test files]
?   	github.com/influxdata/flux/internal/cmd/changelog/cmd	[no test files]
?   	github.com/influxdata/flux/internal/cmd/refactortests	[no test files]
ok  	github.com/influxdata/flux/internal/cmd/refactortests/cmd	5.281s
?   	github.com/influxdata/flux/internal/gen	[no test files]
ok  	github.com/influxdata/flux/internal/parser	1.174s
?   	github.com/influxdata/flux/internal/pkg/syncutil	[no test files]
ok  	github.com/influxdata/flux/internal/scanner	1.097s
ok  	github.com/influxdata/flux/internal/token	1.078s
ok  	github.com/influxdata/flux/interpreter	1.096s
?   	github.com/influxdata/flux/interpreter/interptest	[no test files]
?   	github.com/influxdata/flux/iocounter	[no test files]
ok  	github.com/influxdata/flux/lang	5.005s
ok  	github.com/influxdata/flux/line	1.014s
ok  	github.com/influxdata/flux/memory	1.013s
?   	github.com/influxdata/flux/mock	[no test files]
ok  	github.com/influxdata/flux/parser	1.026s
ok  	github.com/influxdata/flux/plan	5.426s
?   	github.com/influxdata/flux/plan/plantest	[no test files]
?   	github.com/influxdata/flux/querytest	[no test files]
?   	github.com/influxdata/flux/repl	[no test files]
ok  	github.com/influxdata/flux/semantic	1.119s
?   	github.com/influxdata/flux/semantic/semantictest	[no test files]
ok  	github.com/influxdata/flux/stdlib	15.087s
ok  	github.com/influxdata/flux/stdlib/csv	5.451s
?   	github.com/influxdata/flux/stdlib/generate	[no test files]
ok  	github.com/influxdata/flux/stdlib/http	5.094s
ok  	github.com/influxdata/flux/stdlib/influxdata/influxdb	5.115s
ok  	github.com/influxdata/flux/stdlib/influxdata/influxdb/v1	5.039s
ok  	github.com/influxdata/flux/stdlib/kafka	5.113s
ok  	github.com/influxdata/flux/stdlib/math	1.073s
==================
WARNING: DATA RACE
Write at 0x00c00059e908 by goroutine 46:
  github.com/influxdata/flux/values.(*object).updateTypes()
      /root/project/values/object.go:71 +0x5b3
  github.com/influxdata/flux/values.(*object).PolyType()
      /root/project/values/object.go:82 +0x38
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval.func1()
      /root/project/interpreter/interpreter.go:39 +0x72
  github.com/influxdata/flux/values.(*object).Range()
      /root/project/values/object.go:100 +0x116
  github.com/influxdata/flux/interpreter.(*Package).Range()
      /root/project/interpreter/package.go:65 +0x5f
  github.com/influxdata/flux.(*scopeSet).LocalRange()
      /root/project/compile.go:246 +0x8b
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:36 +0x7d
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c00059e908 by goroutine 45:
  github.com/influxdata/flux/values.(*object).updateTypes()
      /root/project/values/object.go:71 +0x5b3
  github.com/influxdata/flux/values.(*object).PolyType()
      /root/project/values/object.go:82 +0x38
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval.func1()
      /root/project/interpreter/interpreter.go:39 +0x72
  github.com/influxdata/flux/values.(*object).Range()
      /root/project/values/object.go:100 +0x116
  github.com/influxdata/flux/interpreter.(*Package).Range()
      /root/project/interpreter/package.go:65 +0x5f
  github.com/influxdata/flux.(*scopeSet).LocalRange()
      /root/project/compile.go:246 +0x8b
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:36 +0x7d
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 46 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 45 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
==================
==================
WARNING: DATA RACE
Write at 0x00c00059e918 by goroutine 46:
  github.com/influxdata/flux/values.(*object).updateTypes()
      /root/project/values/object.go:72 +0x61d
  github.com/influxdata/flux/values.(*object).PolyType()
      /root/project/values/object.go:82 +0x38
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval.func1()
      /root/project/interpreter/interpreter.go:39 +0x72
  github.com/influxdata/flux/values.(*object).Range()
      /root/project/values/object.go:100 +0x116
  github.com/influxdata/flux/interpreter.(*Package).Range()
      /root/project/interpreter/package.go:65 +0x5f
  github.com/influxdata/flux.(*scopeSet).LocalRange()
      /root/project/compile.go:246 +0x8b
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:36 +0x7d
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c00059e918 by goroutine 45:
  github.com/influxdata/flux/values.(*object).updateTypes()
      /root/project/values/object.go:72 +0x61d
  github.com/influxdata/flux/values.(*object).PolyType()
      /root/project/values/object.go:82 +0x38
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval.func1()
      /root/project/interpreter/interpreter.go:39 +0x72
  github.com/influxdata/flux/values.(*object).Range()
      /root/project/values/object.go:100 +0x116
  github.com/influxdata/flux/interpreter.(*Package).Range()
      /root/project/interpreter/package.go:65 +0x5f
  github.com/influxdata/flux.(*scopeSet).LocalRange()
      /root/project/compile.go:246 +0x8b
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:36 +0x7d
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 46 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 45 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
==================
==================
WARNING: DATA RACE
Write at 0x00c00059e928 by goroutine 46:
  github.com/influxdata/flux/values.(*object).updateTypes()
      /root/project/values/object.go:73 +0x657
  github.com/influxdata/flux/values.(*object).PolyType()
      /root/project/values/object.go:82 +0x38
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval.func1()
      /root/project/interpreter/interpreter.go:39 +0x72
  github.com/influxdata/flux/values.(*object).Range()
      /root/project/values/object.go:100 +0x116
  github.com/influxdata/flux/interpreter.(*Package).Range()
      /root/project/interpreter/package.go:65 +0x5f
  github.com/influxdata/flux.(*scopeSet).LocalRange()
      /root/project/compile.go:246 +0x8b
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:36 +0x7d
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c00059e928 by goroutine 45:
  github.com/influxdata/flux/values.(*object).updateTypes()
      /root/project/values/object.go:73 +0x657
  github.com/influxdata/flux/values.(*object).PolyType()
      /root/project/values/object.go:82 +0x38
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval.func1()
      /root/project/interpreter/interpreter.go:39 +0x72
  github.com/influxdata/flux/values.(*object).Range()
      /root/project/values/object.go:100 +0x116
  github.com/influxdata/flux/interpreter.(*Package).Range()
      /root/project/interpreter/package.go:65 +0x5f
  github.com/influxdata/flux.(*scopeSet).LocalRange()
      /root/project/compile.go:246 +0x8b
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:36 +0x7d
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 46 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 45 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
==================
==================
WARNING: DATA RACE
Read at 0x00c0004f31d0 by goroutine 46:
  runtime.mapiterinit()
      /usr/local/go/src/runtime/map.go:734 +0x0
  github.com/influxdata/flux/semantic.object.freeVars()
      /root/project/semantic/poly_types.go:627 +0x8b
  github.com/influxdata/flux/semantic.(*object).freeVars()
      <autogenerated>:1 +0xe7
  github.com/influxdata/flux/semantic.ConstraintGenerator.freshType()
      /root/project/semantic/constraints.go:416 +0x65
  github.com/influxdata/flux/semantic.ConstraintGenerator.typeof()
      /root/project/semantic/constraints.go:144 +0x85bc
  github.com/influxdata/flux/semantic.ConstraintGenerator.Done()
      /root/project/semantic/constraints.go:71 +0x19e
  github.com/influxdata/flux/semantic.(*ConstraintGenerator).Done()
      <autogenerated>:1 +0xbb
  github.com/influxdata/flux/semantic.ScopedVisitor.Done()
      /root/project/semantic/walk.go:389 +0x52
  github.com/influxdata/flux/semantic.(*ScopedVisitor).Done()
      <autogenerated>:1 +0x7d
  github.com/influxdata/flux/semantic.walk()
      /root/project/semantic/walk.go:354 +0x214
  github.com/influxdata/flux/semantic.walk()
      /root/project/semantic/walk.go:79 +0x111b
  github.com/influxdata/flux/semantic.Walk()
      /root/project/semantic/walk.go:6 +0x56
  github.com/influxdata/flux/semantic.GenerateConstraints()
      /root/project/semantic/constraints.go:26 +0x2fc
  github.com/influxdata/flux/semantic.InferTypes()
      /root/project/semantic/inference.go:6 +0x8a
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:45 +0x367
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c0004f31d0 by goroutine 47:
  [failed to restore the stack]

Goroutine 46 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 47 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
==================
==================
WARNING: DATA RACE
Read at 0x00c00027e588 by goroutine 46:
  github.com/influxdata/flux/semantic.object.freeVars()
      /root/project/semantic/poly_types.go:627 +0xae
  github.com/influxdata/flux/semantic.(*object).freeVars()
      <autogenerated>:1 +0xe7
  github.com/influxdata/flux/semantic.ConstraintGenerator.freshType()
      /root/project/semantic/constraints.go:416 +0x65
  github.com/influxdata/flux/semantic.ConstraintGenerator.typeof()
      /root/project/semantic/constraints.go:144 +0x85bc
  github.com/influxdata/flux/semantic.ConstraintGenerator.Done()
      /root/project/semantic/constraints.go:71 +0x19e
  github.com/influxdata/flux/semantic.(*ConstraintGenerator).Done()
      <autogenerated>:1 +0xbb
  github.com/influxdata/flux/semantic.ScopedVisitor.Done()
      /root/project/semantic/walk.go:389 +0x52
  github.com/influxdata/flux/semantic.(*ScopedVisitor).Done()
      <autogenerated>:1 +0x7d
  github.com/influxdata/flux/semantic.walk()
      /root/project/semantic/walk.go:354 +0x214
  github.com/influxdata/flux/semantic.walk()
      /root/project/semantic/walk.go:79 +0x111b
  github.com/influxdata/flux/semantic.Walk()
      /root/project/semantic/walk.go:6 +0x56
  github.com/influxdata/flux/semantic.GenerateConstraints()
      /root/project/semantic/constraints.go:26 +0x2fc
  github.com/influxdata/flux/semantic.InferTypes()
      /root/project/semantic/inference.go:6 +0x8a
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:45 +0x367
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c00027e588 by goroutine 47:
  [failed to restore the stack]

Goroutine 46 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 47 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
==================
==================
WARNING: DATA RACE
Read at 0x00c0001804a0 by goroutine 46:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:221 +0x0
  github.com/influxdata/flux/semantic.ObjectKind.substituteKind()
      /root/project/semantic/label_set.go:124 +0x2be
  github.com/influxdata/flux/semantic.(*ObjectKind).substituteKind()
      <autogenerated>:1 +0x101
  github.com/influxdata/flux/semantic.Substitution.ApplyKind()
      /root/project/semantic/substitution.go:27 +0x115
  github.com/influxdata/flux/semantic.(*Solution).solve()
      /root/project/semantic/solve.go:94 +0x7de
  github.com/influxdata/flux/semantic.SolveConstraints()
      /root/project/semantic/solve.go:15 +0x9e
  github.com/influxdata/flux/semantic.InferTypes()
      /root/project/semantic/inference.go:10 +0xe8
  github.com/influxdata/flux/interpreter.(*Interpreter).Eval()
      /root/project/interpreter/interpreter.go:45 +0x367
  github.com/influxdata/flux.EvalAST()
      /root/project/compile.go:126 +0x381
  github.com/influxdata/flux.CompileAST()
      /root/project/compile.go:63 +0x1fa
  github.com/influxdata/flux.Compile()
      /root/project/compile.go:50 +0x11d
  github.com/influxdata/flux/querytest.NewQueryTestHelper()
      /root/project/querytest/compile.go:37 +0x145
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery.func1()
      /root/project/stdlib/socket/from_test.go:85 +0x93
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c0001804a0 by goroutine 47:
  [failed to restore the stack]

Goroutine 46 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 47 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  github.com/influxdata/flux/stdlib/socket_test.TestFromSocket_NewQuery()
      /root/project/stdlib/socket/from_test.go:83 +0x7d2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
==================
--- FAIL: TestFromSocket_NewQuery (0.00s)
    --- FAIL: TestFromSocket_NewQuery/from_no_args (0.06s)
        testing.go:771: race detected during execution of test
    --- FAIL: TestFromSocket_NewQuery/from_wrong_decoder (0.07s)
        testing.go:771: race detected during execution of test
    --- FAIL: TestFromSocket_NewQuery/from_ok (0.07s)
        testing.go:771: race detected during execution of test
FAIL
FAIL	github.com/influxdata/flux/stdlib/socket	4.092s
?   	github.com/influxdata/flux/stdlib/sql	[no test files]
ok  	github.com/influxdata/flux/stdlib/strings	1.025s
?   	github.com/influxdata/flux/stdlib/system	[no test files]
ok  	github.com/influxdata/flux/stdlib/testing	1.033s
ok  	github.com/influxdata/flux/stdlib/universe	16.889s
ok  	github.com/influxdata/flux/values	1.054s
Makefile:64: recipe for target 'test-race' failed
make: *** [test-race] Error 1
Exited with code 2
```

It appears there's non-synchronized access to the same scope object. Not sure why this only occurs with the socket source.

--- a/complete/complete_test.go
+++ b/complete/complete_test.go
@@ -45,8 +45,10 @@ func TestValue(t *testing.T) {
 func TestFunctionNames(t *testing.T) {
 	boom := values.NewFunction(
 		"boom",
-		semantic.NewFunctionPolyType(semantic.FunctionPolySignature{}),
-		func(values.Object) (values.Value, error) { return nil, nil },
+		semantic.NewFunctionPolyType(semantic.FunctionPolySignature{
+			Return: semantic.Int,
+		}),
+		func(values.Object) (values.Value, error) { return values.NewInt(5), nil },
 		false,
 	)
 	s := interpreter.NewScope()
@@ -72,8 +74,9 @@ func TestFunctionSuggestion(t *testing.T) {
 				"start": semantic.Time,
 				"stop":  semantic.Time,
 			},
+			Return: semantic.Int,
 		}),
-		func(values.Object) (values.Value, error) { return nil, nil },
+		func(values.Object) (values.Value, error) { return values.NewInt(5), nil },
 		false,
 	)
 	s := interpreter.NewScope()

--- a/values/object.go
+++ b/values/object.go
@@ -17,22 +17,49 @@ type Object interface {
 }
 
 type object struct {
-	values map[string]Value
-	poly   semantic.PolyType
-	typ    semantic.Type
-	mod    bool
+	labels semantic.LabelSet
+	values []Value
+	ptyp   map[string]semantic.PolyType
+	mtyp   map[string]semantic.Type
 }
 
 func NewObject() *object {
-	return &object{values: map[string]Value{}}
+	return &object{
+		ptyp: make(map[string]semantic.PolyType),
+		mtyp: make(map[string]semantic.Type),
+	}
 }
-func NewObjectWithValues(values map[string]Value) *object {
-	obj := &object{values: values, mod: true}
-	obj.updateTypes()
-	return obj
+func NewObjectWithValues(vals map[string]Value) *object {
+	l := len(vals)
+
+	labels := make(semantic.LabelSet, 0, l)
+	values := make([]Value, 0, l)
+
+	ptyp := make(map[string]semantic.PolyType, l)
+	mtyp := make(map[string]semantic.Type, l)
+
+	for k, v := range vals {
+		labels = append(labels, k)
+		values = append(values, v)
+
+		ptyp[k] = v.PolyType()
+		mtyp[k] = v.Type()
+	}
+
+	return &object{
+		labels: labels,
+		values: values,
+		ptyp:   ptyp,
+		mtyp:   mtyp,
+	}
 }
 func NewObjectWithBacking(size int) *object {
-	return &object{values: make(map[string]Value, size)}
+	return &object{
+		labels: make(semantic.LabelSet, 0, size),
+		values: make([]Value, 0, size),
+		ptyp:   make(map[string]semantic.PolyType, size),
+		mtyp:   make(map[string]semantic.Type, size),
+	}
 }
 
 func (o *object) IsNull() bool {
@@ -55,49 +82,43 @@ func (o *object) String() string {
 	return b.String()
 }
 
-func (o *object) updateTypes() {
-	if !o.mod {
-		return
-	}
-	l := len(o.values)
-	ts := make(map[string]semantic.Type, l)
-	ps := make(map[string]semantic.PolyType, l)
-	ls := make(semantic.LabelSet, 0, l)
-	for k, v := range o.values {
-		ts[k] = v.Type()
-		ps[k] = v.PolyType()
-		ls = append(ls, k)
-	}
-	o.poly = semantic.NewObjectPolyType(ps, nil, ls)
-	o.typ = semantic.NewObjectType(ts)
-	o.mod = false
-}
-
 func (o *object) Type() semantic.Type {
-	o.updateTypes()
-	return o.typ
+	return semantic.NewObjectType(o.mtyp)
 }
 
 func (o *object) PolyType() semantic.PolyType {
-	o.updateTypes()
-	return o.poly
+	return semantic.NewObjectPolyType(o.ptyp, nil, o.labels)
 }
 
-func (o *object) Set(name string, v Value) {
-	o.values[name] = v
-	o.mod = true
+func (o *object) Set(k string, v Value) {
+	for i, l := range o.labels {
+		if l == k {
+			o.values[i] = v
+			o.ptyp[l] = v.PolyType()
+			o.mtyp[l] = v.Type()
+			return
+		}
+	}
+	o.labels = append(o.labels, k)
+	o.values = append(o.values, v)
+	o.ptyp[k] = v.PolyType()
+	o.mtyp[k] = v.Type()
 }
 func (o *object) Get(name string) (Value, bool) {
-	v, ok := o.values[name]
-	return v, ok
+	for i, l := range o.labels {
+		if name == l {
+			return o.values[i], true
+		}
+	}
+	return nil, false
 }
 func (o *object) Len() int {
 	return len(o.values)
 }
 
 func (o *object) Range(f func(name string, v Value)) {
-	for k, v := range o.values {
-		f(k, v)
+	for i, l := range o.labels {
+		f(l, o.values[i])
 	}
 }
 
@@ -142,9 +163,9 @@ func (o *object) Equal(rhs Value) bool {
 	if o.Len() != r.Len() {
 		return false
 	}
-	for k, v := range o.values {
-		val, ok := r.Get(k)
-		if !ok || !v.Equal(val) {
+	for i, l := range o.labels {
+		val, ok := r.Get(l)
+		if !ok && !o.values[i].Equal(val) {
 			return false
 		}
 	}

